Python/optimizer_bytecodes.c

```diff
@@ -589,8 +589,28 @@ dummy_func(void) {
         self = sym_new_not_null(ctx);
     }
 
-    op(_CHECK_FUNCTION_EXACT_ARGS, (callable, self_or_null, unused[oparg] -- callable, self_or_null, unused[oparg])) {
+    op(_CHECK_FUNCTION_VERSION, (func_version/2, callable, self_or_null, unused[oparg] -- callable, self_or_null, unused[oparg])) {
+        (void)self_or_null;
+        if (sym_is_const(callable) && sym_matches_type(callable, &PyFunction_Type)) {
+            assert(PyFunction_Check(sym_get_const(callable)));
+            REPLACE_OP(this_instr, _CHECK_FUNCTION_VERSION_INLINE, 0, func_version);
+            this_instr->operand1 = (uintptr_t)sym_get_const(callable);
+        }
         sym_set_type(callable, &PyFunction_Type);
+    }
+
+    op(_CHECK_FUNCTION_EXACT_ARGS, (callable, self_or_null, unused[oparg] -- callable, self_or_null, unused[oparg])) {
+        assert((this_instr-1)->opcode == _CHECK_FUNCTION_VERSION || (this_instr-1)->opcode == _CHECK_FUNCTION_VERSION_INLINE);
+        if (sym_is_const(callable) && sym_matches_type(callable, &PyFunction_Type)) {
+            if (sym_is_null(self_or_null) || sym_is_not_null(self_or_null)) {
+                PyFunctionObject *func = (PyFunctionObject *)sym_get_const(callable);
+                PyCodeObject *co = (PyCodeObject *)func->func_code;
+                if (co->co_argcount == oparg + !sym_is_null(self_or_null)) {
+                    // Note: this is only valid if CHECK_FUNCTION_VERSION precedes this instruction.
```

Comment:
`_CHECK_FUNCTION_VERSION` does a `PyFunction_Check`, while `_CHECK_FUNCTION_EXACT_ARGS` only asserts it. So we _must_ have a `_CHECK_FUNCTION_VERSION` before ` _CHECK_FUNCTION_EXACT_ARGS`.

Reply:
`_CHECK_FUNCTION_EXACT_ARGS` relies on `callable` being a function. It does not depend on being preceded by `_CHECK_FUNCTION_VERSION`.

In fact, we could add `sym_set_type(callable, &PyFunction_Type)` since we know that `callable` must be a function to have reached here.

Reply:
Yes but `_CHECK_FUNCTION_VERSION` = `PyFunction_Check(callable)`. So it's a good and safe approximation.

Reply:
It is unnecessary and prevents any optimization that would remove the `_CHECK_FUNCTION_VERSION`.
We should not be adding asserts that can fail when there is nothing wrong.

What we could assert is `sym_matches_type(callable, &PyFunction_Type)`.
We can do that because `_CHECK_FUNCTION_VERSION` ensures that, and any optimization that removes `_CHECK_FUNCTION_VERSION` must already have ensured it.

Reply:
Ok I did that instead.